language/vm/vm_runtime/src/code_cache/module_cache.rs

```diff
@@ -294,26 +294,44 @@ impl<'alloc> VMModuleCache<'alloc> {
                     fields,
                 } => {
                     let mut field_types = vec![];
-
-                    for field in module.field_def_range(*field_count, *fields) {
-                        let ty = try_runtime!(self.resolve_signature_token_with_fetcher(
-                            module,
-                            &module.type_signature_at(field.signature).0,
-                            gas_meter,
-                            fetcher
-                        ));
-                        if let Some(t) = ty {
-                            // `field_types` is initally empty, a single element is pushed per loop
-                            // iteration and the number of iterations is bound to the max size of
-                            // `module.field_def_range()`
-                            // MIRAI cannot currently check this bound in terms of `field_count`.
-                            assume!(field_types.len() < usize::max_value());
-                            field_types.push(t);
-                        } else {
-                            return Ok(Ok(None));
+                    match module.field_def_range(*field_count, *fields) {
+                        Ok(field_defs) => {
+                            for field in field_defs {
+                                let ty = try_runtime!(self.resolve_signature_token_with_fetcher(
+                                    module,
+                                    &module.type_signature_at(field.signature).0,
+                                    gas_meter,
+                                    fetcher
+                                ));
+                                if let Some(t) = ty {
+                                    // `field_types` is initally empty, a single element is pushed
+                                    // per loop iteration and the number of iterations is bound to
+                                    // the max size of `module.field_def_range()`.
+                                    // MIRAI cannot currently check this bound in terms of
+                                    // `field_count`.
+                                    assume!(field_types.len() < usize::max_value());
+                                    field_types.push(t);
+                                } else {
+                                    return Ok(Ok(None));
+                                }
+                            }
+                            StructDef::new(field_types)
+                        }
+                        Err(VMStaticViolation::RangeOutOfBounds(
```

Comment:
You can use [map_err](https://doc.rust-lang.org/std/result/enum.Result.html#method.map_err) to change the error type.

Reply:
I'm also uncertain about this; I need to access the fields of the original error so this match seems to be a clean way to do it.